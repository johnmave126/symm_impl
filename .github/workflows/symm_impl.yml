name: Test

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v1

      - name: Install rust toolchain
        uses: actions-rs/toolchain@v1
        with:
          toolchain: stable
          override: true
          default: true

      - name: Restore Cache
        uses: actions/cache@v2
        with:
          path: |
            ~/.cargo/registry
            ~/.cargo/git
            target
          key: test-cargo-${{ hashFiles('**/Cargo.toml') }}

      - name: Quick check
        run: cargo check

      - name: Test
        run: cargo test