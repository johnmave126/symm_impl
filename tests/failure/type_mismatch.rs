use symm_impl::symmetric;

trait T1<Other> {
    fn op_1(&self, other: Other) -> i32;
}

trait T2<Other> {
    fn op_2(self, other: &Other) -> i32;
}

trait T3<Other> {
    fn op_3(&mut self, other: &Other) -> i32;
}

trait T4<Other> {
    fn op_4(&self, other: &mut Other) -> i32;
}

trait T5<Other> {
    fn op_5<'a, 'b>(&'a self, other: &'b Other) -> i32;
}

struct A {
    a: i32,
}

struct B {
    b: i32,
}

#[symmetric]
impl T1<B> for A {
    fn op_1(&self, other: B) -> i32 {
        1
    }
}
#[symmetric]
impl T2<B> for A {
    fn op_2(self, other: &B) -> i32 {
        2
    }
}
#[symmetric]
impl T3<B> for A {
    fn op_3(&mut self, other: &B) -> i32 {
        3
    }
}
#[symmetric]
impl T4<B> for A {
    fn op_4(&self, other: &mut B) -> i32 {
        4
    }
}
#[symmetric]
impl T5<B> for A {
    fn op_5<'a, 'b>(&'a self, other: &'b B) -> i32 {
        5
    }
}

fn main() {}
