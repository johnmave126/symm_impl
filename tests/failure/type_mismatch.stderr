error: expected a reference
  --> $DIR/type_mismatch.rs:33:20
   |
33 |     fn op_1(&self, other: B) -> i32 {
   |                    ^^^^^

error: mismatched mutability
  --> $DIR/type_mismatch.rs:45:24
   |
45 |     fn op_3(&mut self, other: &B) -> i32 {
   |                        ^^^^^

error: mismatched mutability
  --> $DIR/type_mismatch.rs:51:20
   |
51 |     fn op_4(&self, other: &mut B) -> i32 {
   |                    ^^^^^

error: mismatched lifetime
  --> $DIR/type_mismatch.rs:57:31
   |
57 |     fn op_5<'a, 'b>(&'a self, other: &'b B) -> i32 {
   |                               ^^^^^

error[E0053]: method `op_2` has an incompatible type for trait
  --> $DIR/type_mismatch.rs:38:16
   |
8  |     fn op_2(self, other: &Other) -> i32;
   |                          ------ type in trait
...
38 | impl T2<B> for A {
   |                ^ expected `&A`, found struct `A`
   |
   = note: expected fn pointer `fn(B, &A) -> _`
              found fn pointer `fn(B, A) -> _`

error[E0308]: mismatched types
  --> $DIR/type_mismatch.rs:37:1
   |
37 | #[symmetric]
   | ^^^^^^^^^^^^
   | |
   | expected `&B`, found struct `B`
   | help: consider borrowing here: `&#[symmetric]`
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
